### ISSUE TITLE
RTL (Right-to-Left) text direction for some Asian languages such as Arabic, Hebrew, Persian etc.

### ISSUE THREAD

**Issue by okaeiz** [nextcloud/server#31420]
##Issue
We need to use Persian interface and Persian (Farsi) is a RTL language. This feature is not supported in Nextcloud and the interface strings are translated into Persian but its Left-to-Right.
We have also checked Arabic and it was also LTR.

##What we suggest:
We strongly suggest that Nextcloud developers include the text direction feature.
![report](https://user-images.githubusercontent.com/100111649/156519449-46df529f-d6e1-42ce-88ae-415a3c39eea7.JPG)

**Comment by Spartachetto** [comment]
There should be some work already done. I've seen for example #11716

**Comment by tflidd** [comment]
Perhaps a stupid question, you probably would change the alignment, but would you as well "mirror" the whole design?

**Comment by Spartachetto** [comment]
@jancborchardt Sorry to disturb you, but:
- this issue maybe is already solved
- this is a front end issue

**Comment by okaeiz** [comment]
the issue is not solved yet. 
we have tried multiple options to make it right to left; none of which worked.
We modified core css files using RTLCSS (a tool written in javascript to make a left-to-right interface right-to-left.
We searched for any css patch to include in Custom CSS (a theming app in nextcloud) but we didn't find anything.
We need nextcloud developers to help us out.
This cannot be done without your help.

**Comment by mokazemi** [comment]
There are some problems for Right-to-Left languages.
- **Text boxes** should adopt text (or paragraph) direction automatically, based on the writing language. 
- The **total interface** should be mirrored. (Since Some issues are encountered when the interface is LTR but text isn't.)

**Comment by jancborchardt** [comment]
cc @skjnldsv @marcoambrosini @PVince81 @AndyScherzinger since this is quite an overarching front-end issue.

**Comment by AndyScherzinger** [comment]
Not sure when this could be picked up, since we want to push more for the rewrite to vue and also work on accessibility which both couldn't be worked on given the priorities of version 24. So I think that this should be implemented first before improving RTL support.

**Comment by mokazemi** [comment]
Isn't there anything about RTL layout support for Version 25? 
Since this makes Nextcloud not being much usable for these languages. (Only strings translations is not enought)

**Comment by AndyScherzinger** [comment]
> Isn't there anything about RTL layout support for Version 25?

Not that I know of, maybe @PVince81 is aware of something, but my guess it no.

**Comment by bandarrex** [comment]
it is a front-end issue . the direction must rely on the used language to view the text and tables and containers properly.
open developer mode and addup the page code after html: 
` dir="rtl" `
and see part of the experience

**Comment by ahangarha** [comment]
I have tried manually, and I think the implementation of this feature should be straight forward. If there is no objection against using logical properties in CSS (I mean using `text-align: start|end` instead of `left|right`, or `padding-inline-start|end` instead of `padding-left|right` and other similar cases, I should be able to work on this feature.

Perhaps the only issue is with the support for `inset-inline-start|end` for `left|right` properties in absolute positioned elements.

If there be no objection, this approach helps us to avoid creating rtl styles (which brings more headache for maintenance). Then we need to publish some guidelines for developing apps to respect the same styling conventions to follow the direction of the page.

Green light?

**Comment by AndyScherzinger** [comment]
https://github.com/nextcloud/server/issues/31420#issuecomment-1578262933 What do you think @skjnldsv @nickvergessen @juliushaertl @artonge @ChristophWurst - for some front-enders feedback. In general having RTL supported would be awesome, I just lack the technical knowledge if the approach works or where one would expect challenges _and_ side-effects for app-developers

**Comment by ahangarha** [comment]
The good thing is that my native language is Persian (so I know what I am fixing). I also have several contributions to different projects, fixing their bidirectional text support.

I expect the changes to be very limited, so it shouldn't add more complexities to the project.

I document things so they can become a guideline for other developers.

We only need a reliable way to verify the quality of the changes. If we have automated tests, it would be great. In any case, we will make it work. I have some cool plans for the next 3 months. Wish me for making it realized.

**Comment by AndyScherzinger** [comment]
That sounds awesome @ahangarha 🎉 

Also if you have any troubles from a design perspective feel free to loop in people from @nextcloud/designers or @jancborchardt or @nimishavijay specifically.

Really looking forward to have RTL languages supported ✨

**Comment by alimahwer** [comment]
Great news. Any progress on this feat. My mother language is Arabic, so I can help on reviewing the result and QA.

**Comment by ahangarha** [comment]
@alimahwer No progress yet. But it is good to have you for testing

**Comment by okaeiz** [comment]
I can help you with Persian strings.

**Comment by GameO7er** [comment]
So is this solved in [this](https://github.com/nextcloud/forms/pull/1654) pull request?

**Comment by ahangarha** [comment]
> So is this solved in https://github.com/nextcloud/forms/pull/1654 pull request?
No @GameO7er 
That is some minor fix in forms app. This issue is for the NextCloud itself.

**Comment by GameO7er** [comment]
Thank you for your response.
So, considering that I know different people come to this page through search engines, to make the issue clear and transparent, I would say that the problem with the "[NextCloud Forms](https://apps.nextcloud.com/apps/forms)" has been resolved, but the same problem needs to be addressed in other software as well. Software such as [NextCloud Talk](https://apps.nextcloud.com/apps/spreed), [NextCloud Task](https://apps.nextcloud.com/apps/tasks), [NextCloud Deck](https://apps.nextcloud.com/apps/deck), [NexxtCloud Phone Track](https://apps.nextcloud.com/apps/phonetrack), [NextCloud Calendar](https://apps.nextcloud.com/apps/calendar) and [other existing softwares](https://apps.nextcloud.com/).

**Comment by ahangarha** [comment]
A few colleagues and I are working on adding bidirectional text support to NC server.

There are a few blockers. Should I discuss them here?
- Can we drop IE11 support? In particular, [this commit](https://github.com/nextcloud/server/commit/de84e037e022aa659392d7ba412c023e51de46ed) is problematic.
- There are cases where we have explicit values set for showing elements to the left or right. Namely [these](https://github.com/nextcloud/server/blob/1db6947fcdc734142b3cb2338796faa016cf11af/core/css/tooltip.scss#L56). Can we change these to start|end rather than left|right? (in later PR) Does it have any further impact on any other part of the software?

For now, I don't touch these cases. Please let me know how we can resolve these issues.

**Comment by ahangarha** [comment]
Since nextcloud server uses [this browserslist](https://github.com/nextcloud-libraries/browserslist-config/blob/master/browserlist.config.js), then I think we can totally ignore IE11. But still we have Samasung Internet and Opera Mobile.

May I ignore them as well or you want to keep them in the list? These two browsers don't proper support for logical CSS properties.

![image](https://github.com/nextcloud/server/assets/11241315/9897546a-15fc-4512-92b0-0c9ceea39747)

**Comment by jancborchardt** [comment]
Thanks a bunch for looking into this @ahangarha! :)

> - Can we drop IE11 support? In particular, [this commit](https://github.com/nextcloud/server/commit/de84e037e022aa659392d7ba412c023e51de46ed) is problematic.

@AndyScherzinger what do you think? Also @danxuliu could that commit/fix be done in a different way?

> - There are cases where we have explicit values set for showing elements to the left or right. Namely [these](https://github.com/nextcloud/server/blob/1db6947fcdc734142b3cb2338796faa016cf11af/core/css/tooltip.scss#L56). Can we change these to start|end rather than left|right? (in later PR) Does it have any further impact on any other part of the software?

As far as I know, the tooltip library is being phased out in favor of the native browser title/tooltip functionality, so this might soon be obsolete. cc @JuliaKirschenheuter @ShGKme for more info.

**Comment by nickvergessen** [comment]
> Also danxuliu could that commit/fix be done in a different way?

The commit is 4years old, and the person is OOO for another 2.5 months.
But as per above neither tooltip nor IE11 are supported anymore, so just go ahead.

**Comment by ShGKme** [comment]
> As far as I know, the tooltip library is being phased out in favor of the native browser title/tooltip functionality, so this might soon be obsolete. cc @JuliaKirschenheuter @ShGKme for more info

Yep, it is mostly removed. Some places still use it, but we probably will remove it completely.

**Comment by ahangarha** [comment]
Great.
So I won't bother with IE support.
Now the only issue is (almost) Samsung Internet. We have issue with float/clear styles.
Not sure how many of NextCloud users use the web interface on mobile phones and there, use Samasung Internet.

If I get green light for dropping this browser too, I can wrap up and make the PR in 24 hours.

I must say, this is just the beginning of a series of changes all over NextCloud. These changes wont affect LTR languages (visually) but RTL languages will be impacted (since in some apps they see the content in RTL while the text is LTR; the same issue they face with RTL content in LTR layout even now). So we are moving forward but the team should be aligned and synced about the progress we are going to make.

**Comment by jancborchardt** [comment]
> Now the only issue is (almost) Samsung Internet. We have issue with float/clear styles.
> Not sure how many of NextCloud users use the web interface on mobile phones and there, use Samasung Internet.

According to StatCounter, Samsung Internet has a current market share of 4.58% on mobile: https://gs.statcounter.com/browser-market-share/mobile/worldwide
In 3rd position only after Chrome and Safari – this is much too high of a percentage, so we can not drop support for it.

**Comment by ahangarha** [comment]
> Samsung Internet has a current market share of 4.58% on mobile

That is what is concerning me. I wish there be some statistics about NC users who use this browser.

But I think I can find some workaround if NC wants to support this browser.

**Comment by jancborchardt** [comment]
@ahangarha yeah, we don’t do tracking in this regard though. But considering the many people who use Nextcloud either not only self-hosted but also via companies, schools, governments, very common service providers and more → We have to take the Statcounter values and take them as if they are our own. Thus Samsung Internet is not something we can not support, sorry. :( Hope that’s understandable?

**Comment by ahangarha** [comment]
@jancborchardt 
It is totally understandable. Thanks for your insight.

I have already taken that into account while working on https://github.com/nextcloud/server/pull/42471, and I hope I can get some response in that regard and move forward.

**Comment by AndyScherzinger** [comment]
Implemented via https://github.com/nextcloud/server/pull/47343 and https://github.com/nextcloud/server/pull/47349

**Comment by AndyScherzinger** [comment]
So anyone can try out the latest master and report back any issues you find that still need fixing/polishing.

Some might be here in the server some might be in libraries or apps. Don't worry if unclear, we will take care of transferring the issue to the matching repo in case it is needed 👍

**Comment by cod3r0k** [comment]
Hi, do we have any official update for RTL support (Arabic langunage)?

**Comment by AndyScherzinger** [comment]
Hi @cod3r0k rtl support shipped with v31 m Hub 11 of the Nextcloud server beginning of 2025.
